### ISSUE TITLE
Stream

### ISSUE THREAD

**Issue by vipinvkmenon** [riemann/riemann#1032]
**Describe the bug**

**To Reproduce**
1. Do '...'
2. Change '....'
3. See error

**Expected behavior**

**Screenshots**
If applicable, add configuration, data, output, or screenshots to help explain your problem.

**Background (please complete the following information):**
 - OS: [e.g. Linux]
 - Java/JVM version [e.g. 11]
 - Riemann version [e.g. 0.34]

**Additional context**

**Comment by sanel** [comment]
If you got you correctly, at 192 GB,  riemann used 100% of CPU. However, when you resized the box to 256 GB, riemann had a slow startup and extreme CPU usage? This sounds like a GC issue. Try tuning options, look for `-XX:MaxGCPauseMillis` or `-XX:GCTimeRatio.

If this happens to be a problem, try switching to ZGC. However, be aware it needs way more heap than ParallelGC.

Also, try splitting the load between multiple riemann instances, as explained in https://github.com/riemann/riemann/issues/1003#issuecomment-1012064069. I doubt you'll be able to scale it vertically indefinitely.

**Comment by vipinvkmenon** [comment]
On that note one more question that strikes me is why is the jvm limiting the heap size here despite of having a large amount of headroom..🤔

**Comment by sanel** [comment]
> On that note, if it really is GC ( which is probably the reason) one more question that strikes me is why is the jvm limiting the heap size here despite of having a large amount of headroom

Unless you explicitly set `Xmx`, ParallelGC will use 1/4 of max memory by default. AFAIK, usual practice is to have Xms == Xmx size to get predictable behavior - if they are different, JVM will try to find optimal balance. Probably the best way is to see [official guide](https://docs.oracle.com/en/java/javase/11/gctuning/parallel-collector1.html).

> Wht could be the reason for CPU to sit at 100 even when I increase CPU core count from 48 to 64 though?

You mentioned in the first post you already monitoring stream and netty queue size; try collecting JMX metrics from Riemann or attach tools like VisualVM. There you'll see how much GC is in use. I think ParallelGC will use N or N-1 threads (where N == number of cores) by default. To be sure, check it with:

```sh
java -XX:+PrintFlagsFinal 2>&1 | grep ParallelGCThreads
```

Since you already have VM at 100% of CPU usage for a while, I'd review Riemann configuration and streaming rules first. Heavy use of `by`, `rollup`, putting everyting in index with large/infinite TTL or reinjecting events frequently will contribute to heavy memory/gc/cpu usage and slow streams. If you absolutely must do these things, try to split between e.g. 2-3 Riemann nodes and have one node dedicated for slow/heavy streams only.

**Comment by vipinvkmenon** [comment]
Since the Riemann rules are maintained by several teams and there are 100s of services that rely on this stack items like `by` are used heavily.
We will check into seeing the impact of XMX values, and most likely look into how we can split.
One approach was to split on the basis of metrics coming in based on tcp-server vs graphite-server....or the other approach is to give separate nodes for those that send large amounts of metrics.

**Comment by sanel** [comment]
In that case, I'm not sure how this be considered as a bug on Riemann side; it is hard to reproduce without exact setup you are using.

**Comment by pyr** [comment]
I agree with @sanel here. One thing that could help is moving up in JDK versions and trying the newer GC options.

**Comment by vipinvkmenon** [comment]
Updates:
The run parameters is as follows:
```
    MAX_HEAP=$(awk '/MemTotal/ { printf "%.0f", $2/1024*0.75 }' /proc/meminfo)
    HALF_MAX_HEAP=`expr $MAX_HEAP / 2`
    exec chpst -u vcap:vcap java \
      -XX:+UseParallelGC \
      -XX:+ExitOnOutOfMemoryError \
      -Xms"$HALF_MAX_HEAP"m \
      -Xmx"$MAX_HEAP"m \
      -Djava.io.tmpdir="$TMP_DIR" \
      -jar /var/vcap/packages/riemann/riemann.jar \
      "$CONFIG_DIR"/clojure/riemann.config \
      1>>"$LOG_DIR"/"$JOB_NAME".stdout.log \
      2>>"$LOG_DIR"/"$JOB_NAME".stderr.log &
```
JVM
```
./jcmd "$(ps -C java -o pid=)" VM.version
37825:
OpenJDK 64-Bit Server VM version 17.0.3.0.1+7-LTS
JDK 17.0.3
```
Currently, we are on an `m5.12xlarge` machine. The system is stable with CPU running around 30% to 40% generally and 50%~60% during `busy hours`. On our 192 GB machine, our memory consumption is between 45GB & 70GB (which means memory consumption is less than even 50% at peak). So the current situation is not a concern.
------
Why 'Parallel GC'?

Initially, Riemann was on JVM8 and used CMS
```
      -XX:+CMSParallelRemarkEnabled \
      -XX:+UseCompressedOops \
      -XX:+CMSClassUnloadingEnabled \
```
However, on moving to JDK13, G1GC didn't seem to help as it was unstable and memory kept collecting after a few days and kept crashing. ZGC also didn't seem to help. (and JDK 17 after that). Parallel GC without its throughput collector was the most stable. (We probably would go back to see how GC can be changed to a better option than stop-the-world).

However, from the stack traces, it didn't seem to be GC. Even switching to `-XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC` a no-op GC CPU was at 100 on the BIG machine. 
---
What is concerning is, the moment with the new JVM also, when we upgrade the VM to a higher VM (m5.16xlarge [64CPU/256gb] or c5n.18xlarge [72cpu/192gb]), Riemann starts to become unresponsive and CPU stays at 100% with load touching 120~150.

<img width="714" alt="image" src="https://user-images.githubusercontent.com/3639797/209422846-3f49af13-d4d7-4cde-a555-43716fed5d83.png">

In the given screenshot, The VM was recreated to try the bigger VM size. The Yellow and green parts of the graph were the smaller 48CPU machines in the Blue region when we tried to use the higher configuration 72CPU machine.

So this issue of high CPU only exists on the BIG VM(72 or 64 CPU). Switching back to the smaller(48 CPU) machine brings everything back to normal and stable.

The slight tapering towards the end of the blue region is when we tried to remove rules from the `riemann.config` to see its effect. While it did reduce a bit the application is still playing around 100%, we tried removing the downstream sending of metrics to influxdb assuming that might be a bottleneck... and again it did reduce a bit but still sitting around 100.
-----
I am noting the 2 thread dumps.
[thread_dump1](https://fastthread.io/my-thread-report.jsp?p=c2hhcmVkLzIwMjIvMTIvMjQvdGhyZWFkX2R1bXBfNF80NnBtLnRkdW1wLS01LTUzLTQ4&)

[thread_dump2](https://fastthread.io/my-thread-report.jsp?p=c2hhcmVkLzIwMjIvMTIvMjQvdGhyZWFkX2R1bXBfNF81M19wbS50ZHVtcC0tNS00MC0zOQ==&)

Seems like most threads are parked.
-----

**Comment by sanel** [comment]
Sadly, it is still hard to figure out what is happening without seeing the full `riemann.config` and accompanying files. Of course, seeing a lot of waiting threads means something in the pipeline needs to be corrected, but from the thread dumps, I don't see anything particularly related to Riemann itself (and I might miss it).

I'd try these things in the given order:

 1. Try with a minimal `riemann.config` as possible, like `(streams index)`. Measure it. I'm expecting Riemann not to use 100% of CPU here, and if it does, something might need to be fixed with JVM settings.
 2. Add direct influxdb storage e.g. `(streams index influx)`. Influx can be a bottleneck if it is not configured properly. Check if Influx is running CQs or heavy queries while Riemann is writing to it. Plenty of space to optimize here. Depending on how your streams are organized, Riemann *might wait* until Influx write all the metrics it receives.
 3. Add batched influxdb storage. `(batch)` is a beast on its own, so again, try, measure, rinse and repeat. `(batch)` should reduce pressure on Influx, hopefully improving overall CPU usage.
 4. Add your first custom Riemann streaming. Measure. Proceed with the next custom Riemann streaming function, etc.

I like [riemann-configuration-example](https://github.com/mcorbin/riemann-configuration-example) from @mcorbin  and I find it easy to scale and debug. Try reorganizing `riemann.config` to something like this, if possible:

```clojure
(let [index (index)]
 (streams
   index
   influx
   cpu-check-stream
   mem-check-stream
   ;; disk-check-stream     ;; disk stream is disabled
   custom-stream
   ))
```

and simply commenting/disabling specific streams, you can quickly see how Riemann behaves.
